### ISSUE TITLE
Running a C++ application on Graphene SGX

### ISSUE THREAD

**Issue by gokhalesushant65** [oscarlab/graphene#476]
Hi,
I have a custom c++ application (at the moment just printing some statement on std output).

Trying to run the following way(as per the [official documentation](https://github.com/oscarlab/graphene)  ) with the error as follows:

> (graphene) dell@dell:~/graphene/LibOS/shim/test/native$ ~/graphene/Runtime/pal_loader ./mytest 
/home/dell/graphene/Runtime/pal-Linux is not built, or security mode is not supported

For this, had a look into following previous issues:
#43 , #195 

Regarding issue #43 ,
following is output of the "Runtime" directory:
[graphene_Runtime_contents.txt](https://github.com/oscarlab/graphene/files/2871718/graphene_Runtime_contents.txt)
So, I have pal-linux built correctly.

Regarding #195,
I am confused as to which libraries to add in the manifest. I actually have multiple libraries at the location
> (base) dell@dell:~/graphene/LibOS/shim/test/native$ ls /usr/lib/gcc/x86_64-linux-gnu/5.4.0

Now, I need clarification regarding two points:
1. Libraries at which locations need to be added? I don't know exactly which libraries correspond to libgcc and libstdc++
2. Which files at these locations need to be mentioned in the manifest file?

So, summing up, I need clarifications regarding following two points:
1. Is the syntax of executing the c++ application correct as mentioned at the top of the issue?
2. Libraries to be mentioned in the manifest.

**Comment by dimakuv** [comment]
1. The syntax is incorrect. Notice how the error message complains about `pal-Linux` in your `Runtime/` directory. The output of your `Runtime/` directory shows that you don't have this file but instead have `pal-Linux-SGX`. This makes sense because you have built and want to run Graphene with SGX backend, not with just a Linux backend.

So, you simply need to add `SGX=1`. This environment variable tells the `pal_loader` script to search for and execute `pal-Linux-SGX`. Here is the correct command-line command:

```
$ SGX=1 ~/graphene/Runtime/pal_loader ./mytest
```

2. The libraries to be mentioned in the manifest can be easily found by doing `ldd ./mytest`. This utility lists all library dependencies of your C++ application, with complete paths and versions. Based on this output, you can determine which libraries need to be added in the manifest.

Also, when you run your application under Graphene-SGX, it will fail with error indicating which library should be added to the manifest.

**Comment by gokhalesushant65** [comment]
Thanks dimakuv :+1: 

Now, I have following manifest file(with file extension changed for uploading here):
[mytest_manifest.txt](https://github.com/oscarlab/graphene/files/2872998/mytest_manifest.txt)
Used the issue #19 while writing the manifest file.

I get the following error while trying to run the application:

> (graphene) dell@dell:~/graphene/LibOS/shim/test/apps/mytest$ SGX=1 ./pal_loader ./mytest
file:mytest: error while loading shared libraries: libstdc++.so.6: cannot open shared object file: No such file or directory

The file is actually a link as can be seen from the following output:

> (graphene) dell@dell:~/graphene/LibOS/shim/test/apps/mytest$ ls -l /usr/lib/x86_64-linux-gnu/libstdc++.so.6
lrwxrwxrwx 1 root root 19 Jun 22  2018 /usr/lib/x86_64-linux-gnu/libstdc++.so.6 -> libstdc++.so.6.0.21

So, whats causing the error? Is it the link? or the permissions issue? 

How should I change the manifest file? 



Also, tell me whats the difference between the following declarations?

> loader.env.LD_LIBRARY_PATH = /lib:/lib64

AND 

> fs.mount.lib.type = chroot
fs.mount.lib.path = /lib
fs.mount.lib.uri = file:$(LIBCDIR)

I checked some of the sample manifest files and /lib64 is included in LD_LIBRARY_PATH but not as filesystem to be mounted.


Summarising, need clarifications on 2 points:
1. Changes in the manifest file to be made to resolve the error
2. Difference between "LD_LIBRARY_PATH" and "fs.mount" declarations (i.e. which should be used 
    when)

**Comment by dimakuv** [comment]
1. I think you're almost there. Just try this line:

```
loader.env.LD_LIBRARY_PATH = /lib:/lib/x86_64-linux-gnu:/usr/lib:/usr/lib/x86_64-linux-gnu
```

Now a bit of background. Generally, Graphene follows the standard Linux practices.

* Graphene resolves links to files without problems. You don't need to specify a particular hard link, Graphene will follow a soft link to the actual inode.

* `loader.env.<ENV_VAR_NAME>` is a directive in the manifest which simply overwrites any environment variable from your Linux environment and passes to your application. So, `loader.env.LD_LIBRARY_PATH` overwrites the original Linux variable and passes to the loader+application running under Graphene. This variable lists a set of paths where shared-library files `*.so` should be found (full path, not just prefix!). It is important to specify this variable in the manifest because it has [nasty security implications](https://www.hpc.dtu.dk/?page_id=1180).

* `fs.mount` declarations are very similar to [chroot jails](https://unix.stackexchange.com/questions/105/chroot-jail-what-is-it-and-how-do-i-use-it) in Linux. When you start your application under Graphene, it has zero access to the file system. You need to explicitly whitelist the directories your application can/will access. For this reason, you need at least one `fs.mount`: the one which exposes the real-fs path `file:$(LIBCDIR)` (which resolves to `your-graphene-path/Runtime/`) as `/lib`. In your C++ case, you also need to expose the real-fs path `/usr/` as `/usr/`. In this case, all subdirectories of `/usr/` are visible to Graphene, and `/usr/lib/x86_64-linux-gnu/` in `LD_LIBRARY_PATH` becomes searchable for shared-lib files.

Alternatively, you could change your manifest to be more tight:

```
...
loader.env.LD_LIBRARY_PATH = /lib:/mycpplib

fs.mount.lib.type = chroot
fs.mount.lib.path = /lib
fs.mount.lib.uri = file:$(LIBCDIR)

fs.mount.usr.type = chroot
fs.mount.usr.path = /mycpplib
fs.mount.usr.uri = file:/usr/lib/x86_64-linux-gnu/

sgx.trusted_files.libstdcpp = file:/usr/lib/x86_64-linux-gnu/libstdc++.so.6
...
```

**Comment by gokhalesushant65** [comment]
Thanks dimakuv for the detailed information :+1: 

Was able to run the application. Closing the issue.
